File: LC86.go
// TODO 失败
package main

import (
	"fmt"
)

/**
* Definition for singly-linked list.
* type ListNode struct {
*     Val int
*     Next *ListNode
* }
*/
func partition(head *ListNode, x int) *ListNode {
	less:=new(ListNode)
	lessPos:=less
	more:=new(ListNode)
	morePos:=more
	res:=new(ListNode)
	resPos:=res
	for head!=nil{
		temp:=head.Next
		if head.Val < x{
			lessPos.Next = head
			lessPos=lessPos.Next
			lessPos.Next = nil
		}else{
			morePos.Next = head
			morePos=morePos.Next
			morePos.Next = nil
		}
		head=temp
	}
	//fmt.Print("get\n")
	for less!=nil{
		//fmt.Print("???\n")
		if less.Next!=nil{
			resPos.Next = less.Next
			resPos=resPos.Next
			less=less.Next
		}else{
			break
		}
	}
	for more!=nil{
		//fmt.Print("ccc\n")
		if more.Next!=nil{
			resPos.Next = more.Next
			resPos=resPos.Next
			more=more.Next
		}else{
			break
		}
	}
	return res.Next
}

File: LC86.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode small = new ListNode(-1);
        ListNode smallHead = small;
        ListNode big = new ListNode(-1);
        ListNode bigHead = big;
        while (head != null) {
//            System.out.println(head.val);
            if (head.val < x) {
                small.next = head;
                small = small.next;
            } else {
                big.next = head;
                big = big.next;
            }
            head = head.next;
        }
        // 为什么一定要加这一句？因为 原来big后面可能跟的是个小的 不一定是个null
        big.next = null;
        small.next = bigHead.next;
        return smallHead.next;
    }
}
